### PR TITLE
Cybersource Gateway Updates

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

- Added Cybersource ObsidianHostedGateway changes.
- Added new Javascript.Obsidian project for building Obsidian controls, specifically for the Gateway for now.

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Added Cybersource ObsidianHostedGateway changes.

---------

##### Who reported, requested, or paid for the change?

Island

---------

##### Does this update or add options to the block UI?

Screenshot_Cybersource_RegistrationPayment

Updated to include address:   
image

New generic error message added (will still log the exception with full error):   
image
(Valid test credit card, but Gateway does not accept Discover cards.)

(Actual exception isn't very helpful in this context...)
image

---------

##### What files does it affect?

- .gitignore
- rocks.kfs.CyberSource/Configuration.cs
- rocks.kfs.CyberSource/Controls/CyberSourceHostedPaymentControl.cs
- rocks.kfs.CyberSource/rocks.kfs.CyberSource.csproj
- rocks.kfs.JavaScript.Obsidian/.eslintrc.js
- rocks.kfs.JavaScript.Obsidian/.gitignore
- rocks.kfs.JavaScript.Obsidian/.vscode/launch.json
- rocks.kfs.JavaScript.Obsidian/PROJECT_MIGRATION_LOG.txt
- rocks.kfs.JavaScript.Obsidian/build/.eslintrc.js
- rocks.kfs.JavaScript.Obsidian/build/build-fast.js
- rocks.kfs.JavaScript.Obsidian/build/build-types.js
- rocks.kfs.JavaScript.Obsidian/jest.config.js
- rocks.kfs.JavaScript.Obsidian/nuget.config
- rocks.kfs.JavaScript.Obsidian/package-lock.json
- rocks.kfs.JavaScript.Obsidian/package.json
- rocks.kfs.JavaScript.Obsidian/rocks.kfs.JavaScript.Obsidian.esproj
- rocks.kfs.JavaScript.Obsidian/rocks.kfs.JavaScript.Obsidian.njsproj
- rocks.kfs.JavaScript.Obsidian/rollup.config.js
- rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs
- rocks.kfs.JavaScript.Obsidian/tests/tsconfig.json
- rocks.kfs.JavaScript.Obsidian/tsconfig.json

---------

##### Is it a breaking change for other versions/clients?

There are version 15 and 16 specific versions of the Obsidian project for now. Should not be breaking if on appropriate branch. There is also a compiled block to be included when packaged.

### CODE REVIEW

File: rocks.kfs.CyberSource/Gateway.cs [hunk 1]
```diff
@@ -1421,16 +1432,20 @@ public string CreateCustomerAccount( FinancialGateway financialGateway, Referenc
             Ptsv2paymentsOrderInformationBillTo orderInformationBillTo = new Ptsv2paymentsOrderInformationBillTo(
                 FirstName: paymentInfo.FirstName,
                 LastName: paymentInfo.LastName.IsNullOrWhiteSpace() ? paymentInfo.BusinessName : paymentInfo.LastName,
-                Address1: paymentInfo.Street1,
```
Comment (nateh777) [inline rocks.kfs.CyberSource/Gateway.cs:1424]: Revert these changes

File: rocks.kfs.CyberSource/Gateway.cs [hunk 2]
```diff
@@ -1421,16 +1432,20 @@ public string CreateCustomerAccount( FinancialGateway financialGateway, Referenc
...
+            if ( paymentInfo.Street1.IsNotNullOrWhiteSpace() && paymentInfo.City.IsNotNullOrWhiteSpace() && paymentInfo.State.IsNotNullOrWhiteSpace() && paymentInfo.PostalCode.IsNotNullOrWhiteSpace() )
+            {
+                orderInformationBillTo.Address1 = paymentInfo.Street1;
+                orderInformationBillTo.Address2 = paymentInfo.Street2;
+                orderInformationBillTo.Locality = paymentInfo.City;
+                orderInformationBillTo.AdministrativeArea = paymentInfo.State;
+                orderInformationBillTo.PostalCode = paymentInfo.PostalCode;
+                orderInformationBillTo.Country = paymentInfo.Country;
+            }
+
```
Comment (nateh777) [inline rocks.kfs.CyberSource/Gateway.cs:1448]: same as above

File: rocks.kfs.CyberSource/Model/AddressToken.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Rock.Model;
+
+namespace rocks.kfs.CyberSource.Model
+{
+    public class AddressToken
+    {
+        public Location BillingAddress { get; set; }
```
Comment (nateh777) [inline rocks.kfs.CyberSource/Model/AddressToken.cs:12]: Make note to have Ross test this with his production environment with international addresses and culture.

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 4]
```diff
@@ -191,104 +221,19 @@
...
-  .nmi-payment-inputs .credit-card-exp-input,
-  .nmi-payment-inputs .credit-card-cvv-input {
-    -webkit-box-flex: 0;
-    -ms-flex: 0 0 auto;
-    flex: 0 0 auto;
-    max-width: 100px;
-  }
-}
-`;
+    const standardStyling = ``;
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:235]: Is this used?

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 5]
```diff
@@ -191,104 +221,19 @@
...
-    async function loadCollectJSAsync(tokenizationKey: string): Promise<boolean> {
-        if (typeof CollectJS === "undefined") {
+    async function loadMicroformJsAsync(microFormJWK: string): Promise<boolean> {
+        if (typeof Flex === "undefined") {
             const script = document.createElement("script");
             script.type = "text/javascript";
-            script.src = "https://secure.nmi.com/token/Collect.js";
-            script.setAttribute("data-tokenization-key", tokenizationKey);
+            script.src = props.settings.microFormJsPath;
+            //script.setAttribute("data-tokenization-key", microFormJWK);
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:248]: remove this line

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 6]
```diff
@@ -0,0 +1,727 @@
...
+     * Ensures the Cybersource Flex Microform script is loaded into the browser.
+     *
+     * @param microFormJWK The microform jwk key that will be used to initialize the script.
+     */
+    async function loadMicroformJsAsync(microFormJWK: string): Promise<boolean> {
+        if (typeof Flex === "undefined") {
+            const script = document.createElement("script");
+            script.type = "text/javascript";
+            script.src = props.settings.microFormJsPath;
+            //script.setAttribute("data-tokenization-key", microFormJWK);
```
Comment (guymaness2) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:248]:
```suggestion
```

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 7]
```diff
@@ -403,73 +360,126 @@
...
+                 clearInterval(FlexJS?.loadCheckInterval);
+             }
+
+            validationMessage.value = 'We\'re sorry your session has timed out. Please reload the page to try again.';
+            validationMessageShowReload.value = true;
+
+            var actionBtn = document.querySelector('.btn-give-now, .js-submit-hostedpaymentinfo, .navigation.actions .btn, .registration-entry .actions .btn-primary');
+            if (actionBtn != null) {
+                actionBtn.classList.add('disabled');
+                actionBtn.removeAttribute('href');
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:438]: This can possibly be refined with at least disabled, perhaps you can remove all click events as well

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 8]
```diff
@@ -403,73 +360,126 @@
...
+             expirationYear: ccexpvalue.year
+         };
+
+         FlexJS.inSubmission = true;
+
+         FlexJS?.microform?.createToken(options, function (err, token) {
+             if (err) {
+                 // handle error
+                 console.error(err);
+                 emit(GatewayEmitStrings.Error, err.message);
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:502]: Should we append something to this error message so it isn't just the gateway's issue?

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 9]
```diff
@@ -623,17 +577,17 @@
...
 
         // A timeout callback will fire due to a timeout or incomplete
-        // input fields (CollectJS doesn't tell us why).
+        // input fields (CyberSource doesn't tell us why).
         console.log("The tokenization didn't respond in the expected timeframe. This could be due to an invalid or incomplete field or poor connectivity - " + Date());
 
         // Since we don't know exactly what happened, lets see if it might
         // be invalid inputs by checking them all manually.
-        const validationErrors = validateInputs();
+        //const validationErrors = validateInputs();
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:616]: Should we leave commented out code to come back to this or just remove it as the gateway validates it instead?

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 10]
```diff
@@ -614,7 +568,7 @@
     }
 
     /**
-     * Callback function that handles the timeout scenario of CollectJS.
+     * Callback function that handles the timeout scenario of CyberSource.
      */
     const timeoutCallback: TimeoutCallback = () => {
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:604]: Is this being used? If it is not but you want to keep it at least add a comment as such.

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 11]
```diff
@@ -672,7 +626,7 @@
 
         const validationErrors = validateInputs();
 
-        if (hasAttemptedSubmit && !(CollectJS?.inSubmission ?? false) && !hasReceivedToken) {
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs]: Is this being used? If not add a comment for it.

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 12]
```diff
@@ -205,6 +213,7 @@
         gatewayUrl: string;
         microFormJsPath: string;
         microFormJWK: string;
+        address?: AddressControlBag | null;
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:218]: Remove this line?

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 13]
```diff
@@ -517,6 +534,7 @@
     const tokenResponseSent = ref(false);
 
     let ccexpvalue:MonthYearDay = { month: 0, year: 0, day: 0 };
+    const address = ref<AddressControlBag | undefined>(props.settings.address ?? undefined);
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:550]: Where should I document how I got around the typescript null error? 
In the address control itself I had to add `// @ts-ignore` above each line that generated an exception.

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs [hunk 14]
```diff
@@ -0,0 +1,727 @@
...
+            customStyles,
+            customCss,
+            focusCss,
+            invalidCss,
+            placeholderCss,
+            cardIcons,
+            timeoutDuration: 10000,
+            initialLoadTimeout: 30000,
+            flexTimeout: 900000,
+            flexTimeLoaded: new Date().getTime(),
```
Comment (nateh777) [inline rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:395]: Can this load from props.settings instead? So the FlexTimeLoaded comes from c# instead.

File: rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs (hunk 12 above)
Comment (nateh777) [reply rocks.kfs.JavaScript.Obsidian/src/Controls/cyberSourceGatewayControl.obs:218]: I am using it in the const definition, still not sure if it is absolutely necessary though as this should only be values that come from the c# code.

**Review by guymaness2 — APPROVED** [review]
